src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/EETypeNode.GCDesc.cs

```diff
@@ -0,0 +1,200 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+using Internal.TypeSystem;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    partial class EETypeNode
+    {
+        private int GCDescSize
+        {
+            get
+            {
+                if (!_constructed || _type.IsGenericDefinition)
+                    return 0;
+
+                if (_type.IsArray)
+                {
+                    TypeDesc elementType = ((ArrayType)_type).ElementType;
+                    if (elementType.IsObjRef)
+                    {
+                        // For efficiency this is special cased and encoded as one serie.
+                        return 3 * _type.Context.Target.PointerSize;
+                    }
+                    else
+                    {
+                        if (elementType.IsPointer)
+                        {
+                            return 0;
+                        }
+                        else if (elementType.IsByRef)
+                        {
+                            throw new BadImageFormatException();
+                        }
+                        else
+                        {
+                            var defType = (DefType)elementType;
+                            if (defType.ContainsPointers)
+                            {
+                                int numSeries = GCPointerMap.FromInstanceLayout(defType).NumSeries;
+                                Debug.Assert(numSeries > 0);
+                                return (numSeries + 2) * _type.Context.Target.PointerSize;
+                            }
+                            return 0;
+                        }
+                    }
+                }
+                else
+                {
+                    var defType = (DefType)_type;
+                    if (defType.ContainsPointers)
+                    {
+                        int numSeries = GCPointerMap.FromInstanceLayout(defType).NumSeries;
+                        Debug.Assert(numSeries > 0);
+                        return (numSeries * 2 + 1) * _type.Context.Target.PointerSize;
+                    }
+                    return 0;
+                }
+            }
+        }
+
+        private void OutputGCDesc(ref ObjectDataBuilder builder)
+        {
+            if (!_constructed || _type.IsGenericDefinition)
+            {
+                Debug.Assert(GCDescSize == 0);
+                return;
+            }
+
+            int initialBuilderPosition = builder.CountBytes;
+
+            if (_type.IsArray)
+            {
+                TypeDesc elementType = ((ArrayType)_type).ElementType;
+                if (elementType.IsObjRef)
+                {
+                    OutputStandardGCDesc(ref builder,
+                        new GCPointerMap(new[] { 1 << 2 }, 3),
```

Comment:
Could you please at least structure the code so that enabling this optimization is just about uncommenting a line - and add TODO comment next to it to?